### ISSUE TITLE
BiLSTM and Training Time

### ISSUE THREAD

**Issue by TaoDream** [eragonruan/text-detection-ctpn#12]
Thanks for sharing your implementation with us. I have implemented CTPN with Caffe which failed to converge when adding LSTM. 
First, I want to ask whether you have added the BiLSTM in your code or not. I am new to tensorflow. After looking at the code, I think you just implement the LSTM not the BiLSTM, is it right ?
Second, I want to ask how long did you train your model? I  have run the train script of your programs on a GPU device. It seems that it would take 5-6 days to finish the first 180000 iterations. 

Thanks very much.

**Comment by eragonruan** [comment]
@TaoDream yes, I just implement the LSTM , not BiLSTM, 
I train this model with GTX1070, and it takes about 6 hours with nms written in cuda.
But it may takes a little longer since cython version nms is provided in this repo.

**Comment by TaoDream** [comment]
Thanks very much. Your reply is appreciated. It helps me a lot
Another two question, 
First,  Is your model is trained which used to detect the ID cards  is trained by the 6k pictures you have provided in the good drive ? I have looked at the train data, it doesn't have too much Chinese characters, but it generalizes very well on Chinese characters.

Second, is it sufficient to only compute the rpn_loss_cls and rpn_loss_box ? . I have inferred that from deploy.prototxt of the original repo, I think they have removed the the final rcnn part.

**Comment by eragonruan** [comment]
@TaoDream no, I just use ID card detection as an example to demonstrate the result, the model can be used in any horizontal scene text detection situation. 
Also, the dataset I use is multi-lingual scene text detection dataset, it contains about 800 images in chinese.
There's no rcnn part in the origin paper. A rpn network is sufficient in my experiment.

**Comment by TaoDream** [comment]
Thanks very much.  不好意思，我表达的不是清楚，其实。我是想问，你的最终给我们提供的模型文件，是用google drive上的6k图片训练的吗？我在考虑，需不需要重新训练模型。
你的工作非常优秀，非常感谢你。我自己搞了一周，最后还是没能达到理想效果，你这个实现，雪中送碳。 这应该是，github上第一个完整的CTPN的实现。

**Comment by eragonruan** [comment]
@TaoDream yes, this model is trained with 6k images provided in google drive.

**Comment by TaoDream** [comment]
OK. Thanks very much.
I will dig deeper into your codes.

**Comment by cipri-tom** [comment]
@eragonruan I see there is `nms_kernel.cu` in the utils folder and I set the training flag `USE_GPU_NMS`. I think it works since I don't see very high CPU usage during training.

However, my training is still very slow. Using 1000 images on a GPU 1080 Ti and it took 12 hours to do 20K iterations 😞 

Am I doing something wrong ? Could it be that the `assert`s in the pyFunc slow this down a lot ?

**Comment by eragonruan** [comment]
@cipri-tom did you build the library successfully? it takes me about 0.2s per iter. use 
`watch -n 1 nvidia-smi` to check the GPU-Util， see if gpu usage is low

**Comment by cipri-tom** [comment]
Yes, I built it successfully.

I did exactly that `watch` command and saw GPU usage was very low. I profiled the code and saw a very very long time was spend in `anchor_target_layer_tf`, when calculating the IoU (with the Cython code).

The cause could have been that my images are very big ? after resize, they are all `600 x 850`. Which resulted in some 17K anchors to be generated. Times 400 ground truth mini-boxes on the original image ... There was quite a lot of possible overlaps to compute, all on CPU.

So I lowered the number of anchors generated (because I don't need all heights, I know my text is within a small range). Furthermore, the `argmax(axis=0)` in the same function took considerable time, so I replaced that version with a custom Cython one (see [here](https://github.com/numpy/numpy/issues/9182)). 

Now I am down to 0.2s per iter as well.

**Comment by eragonruan** [comment]
@cipri-tom I think you are right. The training dataset I use only have a few text instance per image, but you got 400, this may be the cause of the low speed.

**Comment by lxx1884896** [comment]
@TaoDream  Hi,I have met the  same question as you .I train the data with a 1080Ti,but it's about 5~6s/iter,it's  so slow, and I want to know how did you solve the problem about speed,thank you .

**Comment by lxx1884896** [comment]
@cipri-tom ，hello,I also use a 1080Ti to train the data, but it's about 5~6s/iter,it's so slow, can you kindly tell me how do you solve the problem about speed, thank you .

**Comment by cipri-tom** [comment]
hi @lxx1884896 

My repo is out of date, but you can see what I did here: https://github.com/cipri-tom/text-detection-ctpn/commits/master 

Essentially, 2 things reduce the time A LOT !

1. Cython argmax -- argmax is not optimal when you're not doing it on the last axis i.e. `argmax(axis=0)` needs to copy all the data before doing the operation. When the matrix is big (which it is) this takes a lot of CPU time. [This commit](https://github.com/cipri-tom/text-detection-ctpn/commit/f0c029a3899425b6d60fb7b5cbaab7c5831864c7) shows the files that I added (ignore the `.c` file, it is generated automatically). 

2. reduce the number of anchors. The original version uses 10 anchors of various sizes. For my application, text had almost always the same height in the image, so I reduced to just 3 anchors which further reduced the dimensionality of the computation. Be careful that there are several places in the code where this is hardcoded 😢 . You can see them in [this commit](https://github.com/cipri-tom/text-detection-ctpn/commit/bf70e8166797637a978c380ed4e93f4591f58dde) where I replaced `10` by `3` and also the list of heights
